src/hotspot/share/runtime/runtimeUpcalls.cpp

```diff
@@ -0,0 +1,204 @@
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/methodDetails.hpp"
+#include "runtime/runtimeUpcalls.hpp"
+
+GrowableArray<RuntimeUpcallInfo*>* RuntimeUpcalls::_upcalls[RuntimeUpcallType::numTypes];
+RuntimeUpcalls::State RuntimeUpcalls::_state = RuntimeUpcalls::Uninitialized;
+
+bool runtimeUpcalls_open_registration() {
+  return RuntimeUpcalls::open_upcall_registration();
+}
+
+bool RuntimeUpcalls::open_upcall_registration() {
+  assert(_state == Uninitialized, "upcalls are already open");
+  for (int i = 0; i < RuntimeUpcallType::numTypes; i++) {
+    RuntimeUpcalls::_upcalls[i] = nullptr;
+  }
+  _state = Open;
+  return true;
+}
+
+void runtimeUpcalls_close_registration() {
+  RuntimeUpcalls::close_upcall_registration();
+}
+
+void RuntimeUpcalls::close_upcall_registration() {
+  assert(_state == Open, "upcalls are not open");
+  _state = Closed;
+}
+
+void RuntimeUpcalls::mark_for_upcalls(RuntimeUpcallType upcallType, const methodHandle& method) {
+  if (_upcalls[upcallType] != nullptr) {
+    MethodDetails md(method);
+    for(RuntimeUpcallInfo* info : *(_upcalls[upcallType])) {
+      if(info->includes(md)) {
+        switch(upcallType) {
+          case onMethodEntry: method->set_has_upcall_on_method_entry(true); break;
+          case onMethodExit:  method->set_has_upcall_on_method_exit(true);  break;
+          default:            ShouldNotReachHere();
+        }
+        break;
+      }
+    }
+  }
+}
+
+bool RuntimeUpcalls::register_upcall(RuntimeUpcallType upcallType, RuntimeUpcallInfo* info) {
+  assert(info != nullptr, "upcall info is null");
+  if (_upcalls[upcallType] == nullptr) {
+    _upcalls[upcallType] = new (mtServiceability) GrowableArray<RuntimeUpcallInfo*>(1, mtServiceability);
+  }
+  info->set_index(_upcalls[upcallType]->length());
+  _upcalls[upcallType]->append(info);
+  return true;
+}
+
+int RuntimeUpcalls::get_num_upcalls(RuntimeUpcallType upcallType) {
+  return (_upcalls[upcallType] == nullptr) ? 0 : _upcalls[upcallType]->length();
+}
+
+void RuntimeUpcalls::upcall_redirect(RuntimeUpcallType upcallType, JavaThread* current, Method* method) {
+  MethodDetails md(method);
+
+  // This redirection occurs when there are more than one upcalls setup.  Currently each method is marked
+  // to indicate either none, entry and/or exit upcalls (two bits total); then we have to iterate over
+  // all upcalls and test the method details to determine which upcalls to call.  This is not optimal.
+  // One possible optimization is to use more bits to support more upcalls.  The method flags currently use 18
+  // out of 32 bits, so there are still 14 bits available for use.  We could set a limit of say 4-8 entry/exit
+  // upcalls combined, leaving 10-6 bits for other uses.  This still requires a redirect here to determine
+  // which upcalls to call, but it would be more efficient than the current implementation as we'd avoid the
+  // method matching and simply map bits to indexes.
+
```

Comment:
Another way to speed it up is to store a hashset of Method pointers in each RuntimeUpcallInfo when `RuntimeUpcalls::mark_for_upcalls` is called for the method. Then the method matching can be replaced with a  lookup in the hashset.